Source/SwiftLintFramework/Rules/GenericTypeNameRule.swift

```diff
@@ -0,0 +1,235 @@
+//
+//  GenericTypeNameRule.swift
+//  SwiftLint
+//
+//  Created by Marcelo Fabri on 12/25/16.
+//  Copyright © 2016 Realm. All rights reserved.
+//
+
+import Foundation
+import SourceKittenFramework
+
+public struct GenericTypeNameRule: ASTRule, ConfigurationProviderRule {
+    public var configuration = NameConfiguration(minLengthWarning: 1,
+                                                 minLengthError: 0,
+                                                 maxLengthWarning: 20,
+                                                 maxLengthError: 1000)
+
+    public init() {}
+
+    public static let description = RuleDescription(
+        identifier: "generic_type_name",
+        name: "Generic Type Name",
+        description: "Generic type name should only contain alphanumeric characters, start with an " +
+                     "uppercase character and span between 1 and 20 characters in length.",
+        nonTriggeringExamples: [
+            "func foo<T>() {}\n",
+            "func foo<T>() -> T {}\n",
+            "func foo<T, U>(param: U) -> T {}\n",
+            "func foo<T: Hashable, U: Rule>(param: U) -> T {}\n",
+            "struct Foo<T> {}\n",
+            "class Foo<T> {}\n",
+            "enum Foo<T> {}\n",
+            "func run(_ options: NoOptions<CommandantError<()>>) {}\n",
+            "func foo(_ options: Set<type>) {}\n",
+            "func < <T: Comparable>(lhs: T?, rhs: T?) -> Bool\n",
+            "func configureWith(data: Either<MessageThread, (project: Project, backing: Backing)>)\n",
+            "typealias StringDictionary<T> = Dictionary<String, T>\n",
+            "typealias BackwardTriple<T1, T2, T3> = (T3, T2, T1)\n",
+            "typealias DictionaryOfStrings<T : Hashable> = Dictionary<T, String>\n"
+        ],
+        triggeringExamples: [
+            "func foo<↓T_Foo>() {}\n",
+            "func foo<T, ↓U_Foo>(param: U_Foo) -> T {}\n",
+            "func foo<↓\(String(repeating: "T", count: 21))>() {}\n",
+            "func foo<↓type>() {}\n",
+            "typealias StringDictionary<↓T_Foo> = Dictionary<String, T_Foo>\n",
+            "typealias BackwardTriple<T1, ↓T2_Bar, T3> = (T3, T2_Bar, T1)\n",
+            "typealias DictionaryOfStrings<↓T_Foo: Hashable> = Dictionary<T, String>\n"
+        ] + ["class", "struct", "enum"].flatMap { type in
+            [
+                "\(type) Foo<↓T_Foo> {}\n",
+                "\(type) Foo<T, ↓U_Foo> {}\n",
+                "\(type) Foo<↓T_Foo, ↓U_Foo> {}\n",
+                "\(type) Foo<↓\(String(repeating: "T", count: 21))> {}\n",
+                "\(type) Foo<↓type> {}\n"
+            ]
+        }
+    )
+
+    private let genericTypePattern = "<(\\s*\\w.*?)>"
+    private var genericTypeRegex: NSRegularExpression {
```

Comment:
Because I wanted to make sure it's read-only and `regex` caches the objects anyway